content/blog/2022/2022-11-21-building-csv-importer-with-nodejs.md

```diff
@@ -0,0 +1,953 @@
+            attributes: ['id', 'name', 'email', 'username', 'avatar'],
+            required: true
+        }],
+        attributes: {
+            exclude: ['managedBy']
+        }
+    })
+      .then((data) => {
+        res.send(data);
+      })
+      .catch((err) => {
+        res.status(500).send({
+          message:
+            err.message
+             || "There might be some error occurred while retrieving employees from database.",
+        });
+      });
+  };
+
+export default getEmployees;
+```
+
+As we can notice, while defining the data model we had a field named `managedBy` which co-relates the Employee with its manager in the same table. Till now, we have defined associations for the One-To-Many relationship, but now to fetch the data we need to opt for either *lazy loading* or *eager loading* concept.
+
+Lazy loading refers to the technique of fetching the related data only when you truly want it. Eager loading, on the other hand, refers to the approach of requesting everything at once, starting from the beginning, with a bigger query. It is a process of simultaneously requesting data from one primary model and one or more associated models. This is a query involving one or more joins at the SQL level. For our use case, we need to opt for the eager-loading concept to map the same model to retrieve child values.
+
+In Sequelize, eager loading is mainly done by using the `include` option on a model finder query (such as `findOne`, `findAll`, etc). Thus, we have defined the following option for our `include`:
+
+```javasc
+include: [{
+    model: Employee,
+    as: 'children',
+    attributes: ['id', 'name', 'email', 'username', 'avatar'],
+    required: true
+}]
+```
+
+* `model` defines the data model that we need to associate.
+* `as` would alias the association column.
+* `attributes` would return the fields to be retrieved for the associated model.
+* `required` if returned false would define `OUTER JOIN`, but if returned true then it will define `INNER JOIN`.
+
+Finally, we have also defined `exclude` to exclude a given attribute from the final result as we don’t want to retrieve `managedBy` attribute since we have defined a `children` attribute now.
+
+```javascript
+attributes: {
+    exclude: ['managedBy']
+}
+```
+
+Next, we need to define the routes for each of these controller methods to map them to API endpoints:
+
+```javascript
+import { Router } from 'express';
+import csvController from '../controllers/csv.controller.js';
+import getEmployees from '../controllers/employee.controller.js';
+import uploadFile from '../middleware/upload.js';
+
+const router = Router();
+
+let routes = (app) => {
+  // CSV
+  router.post('/csv/upload', uploadFile.single('file'), csvController.upload);
+  router.get('/csv/download', csvController.download);
+
+  // Employees
+  router.get('/employees', getEmployees);
+
+  app.use("/api", router);
+};
+
+export default routes;
+```
+
+We have defined three endpoints:
```

Comment:
Maybe move this up to the top before the code examples, so the reader knows what's coming

Reply:
Moved this above the code example.